File: Sprite.java
//package Paradigms.sprites;

import java.awt.Graphics;
// import Paradigms.models.*;

public abstract class Sprite {
  int x;
  int y;
  int prevX;
  int prevY;
  int w;
  int h;
  boolean isMario;

  Sprite() {
    x = 0;
    y = 0;
    w = 0;
    h = 0;
  }

  Sprite(int _x, int _y, int _w, int _h) {
    x = _x;
    y = _y;
    w = _w;
    h = _h;
  }

  Sprite(Sprite s) {
    x = s.x;
    y = s.y;
    w = s.w;
    h = s.h;
    prevX = s.prevX;
    prevY = s.prevY;
  }

  public abstract void update(Model m);
  public abstract void draw(Graphics g, Model m);
  public abstract Json marshall();
  public abstract void spriteHit(Model m, Sprite s);

  boolean isPointInsdie(int _x, int _y) {
    if (x + w > _x &&
				x     < _x &&
				y + h > _y &&
				y     < _y) {
          return true;
    }
    return false;
  }

  boolean isColliding(Sprite s) {
    if (x + w > s.x       &&
				x     < s.x + s.w &&
				y + h > s.y           &&
				y     < s.y + s.h) {
          return true;
    }
    return false;
  }

  int getOut(Sprite s) { // return what direction you hit the brick
    if (y + h >= s.y && !(prevY + h > s.y)) { //from top
      y = s.y - h;
      return 1;
    } else if (y <= s.y + s.h && !(prevY < s.y + s.h)) { // from bottom
      y = s.y + s.h;
      return 3;
    } else if (x + w >= s.x && !(prevX + w > s.x)) { // from left
      x = s.x - w;
      return 4;
    } else if (x <= s.x + s.w && !(prevX < s.x + s.w)) { // from right
      x = s.x + s.w;
      return 2;
    }
    return 0;
  }
}

File: sprite.py
class Sprite:
  def __init__(self, x, y, w, h):
    self.x = x
    self.y = y
    self.w = w
    self.h = h
    self.prevX = x
    self.prevY = y

  def isPointInside(self, x, y):
    return (
      self.x + self.w > x and
      self.x          < x and
      self.y + self.h > y and
      self.y          < y
    )

  def isColliding(self, s):
    return (
      self.x + self.w > s.x       and
      self.x          < s.x + s.w and
      self.y + self.h > s.y       and
      self.y          < s.y + s.h
    )

  def getOut(self, s):
    if self.y + self.h >= s.y and (not (self.prevY + self.h > s.y)): # from top
      self.y = s.y - self.h
      return 1
    elif self.y <= s.y + s.h and (not (self.prevY < s.y + s.h)): # from bottom
      self.y = s.y + s.h
      return 3
    elif self.x + self.w >= s.x and not (self.prevX + self.w > s.x): # from left
      self.x = s.x - self.w
      return 4
    elif self.x <= s.x + s.w and not (self.prevX < s.x + s.w): # from right
      self.x = s.x + s.w
      return 2
    return 0